webrtc.html

```diff
@@ -509,33 +510,54 @@
 
         <ol>
           <li>
-            <p>Append an object representing the current call being handled
-            (i.e. function name and corresponding arguments) to the
-            <code>operations</code> array.</p>
+            <p>Let <var>operation</var> be a function executing the steps of the
+            function in question.</p>
+          </li>
+          <li>
+            <p>If <var>connection</var>'s <a href=
+              "#dom-peerconnection-signaling-state"><code>RTCPeerConnection</code>
+              signalingState</a> is <code>closed</code>, throw an
+              <code>InvalidStateError</code> exception and abort these steps.</p>
+          </li>
+          <li>
+            <p>If <var>connection</var>'s <var>operationsLength</var> is
+            <code>0</code>, let <var>p</var> be the result of executing
+            <var>operation</var>; otherwise, let <var>p</var> be the result of
+            transforming <var>connection</var>'s <var>operations</var> by a
+            fulfillment handler that runs the following steps:</p>
+            <ol>
+              <li>
+                <p>Return the result of executing <var>operation</var>.</p>
+              </li>
+            </ol>
           </li>
 
           <li>
-            <p>If the length of the <code>operations</code> array is exactly 1,
-            execute the function from the front of the queue
-            asynchronously.</p>
+            <p>Increment <var>connection</var>'s <var>operationsLength</var> by
+            <code>1</code>.</p>
+          </li>
+          <li>
+            <p>Set <var>connection</var>'s <var>operations</var> to the result
+            of transforming <var>p</var> by a fulfillment and
+            rejection handler, both of which MUST immediately decrement
+            <var>connection</var>'s <var>operationsLength</var> by
+            <code>1</code> and return <var>undefined</var> (this has no effect
+            on <var>p</var>). To ensure the decrement happens before any other
+            use of <var>p</var>, both handlers MUST not do any thing else.</p>
```

Comment:
On Tue, Sep 22, 2015 at 7:25 PM, jan-ivar <notifications@github.com> wrote:

> In webrtc.html
> <https://github.com/w3c/webrtc-pc/pull/307#discussion_r40165975>:
>
> >            </li>
> >
> >            <li>
> > -            <p>If the length of the <code>operations</code> array is exactly 1,
> > -            execute the function from the front of the queue
> > -            asynchronously.</p>
> > +            <p>Increment <var>connection</var>'s <var>operationsLength</var> by
> > +            <code>1</code>.</p>
> > +          </li>
> > +          <li>
> > +            <p>Set <var>connection</var>'s <var>operations</var> to the result
> > +            of transforming <var>p</var> by a fulfillment and
> > +            rejection handler, both of which MUST immediately decrement
> > +            <var>connection</var>'s <var>operationsLength</var> by
> > +            <code>1</code> and return <var>undefined</var> (this has no effect
> > +            on <var>p</var>). To ensure the decrement happens before any other
> > +            use of <var>p</var>, both handlers MUST not do any thing else.</p>
>
> I'm sad to hear things need to be "impossible" before adopting a standard.
>
Before we throw something very complicated into a standard, we should first
see if it's possible accomplish the same thing with something more simple.​
 You are proposing replacing something simple with something complicated.
I think we should see if we can accomplish the same thing while keeping it
simple.

> I started out "just tweaking", and it led me here when I realized I was
> reinventing .then
> <https://people.mozilla.org/%7Ejorendorff/es6-draft.html#sec-performpromisethen>,
> poorly. [1]
>
> This was 9 months ago when I added this to Firefox
> <https://bugzilla.mozilla.org/show_bug.cgi?id=1106675>.
>
> [1] Promises nail down behaviors and timings around asynchronous
> operations in ways that aren't easily replicable (since a microtask queue
> is used). Also, error handling needs to be done right.
>
Can you please take a look at https://github.com/w3c/webrtc-pc/pull/314,
which I wrote to update the current approach (queue of operations) in terms
of promises?  Can you tell me under what circumstances the behavior would
be different than this PR (chain of promises)?  If there are real behavior
differences, then I'd like to attempt to achieve the same thing with a
queue of operations while keeping it simple.  If there are no behavior
differences, then we can keep the more simple model of a queue of
operations.

​



> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/webrtc-pc/pull/307/files#r40165975>.
>